### PR TITLE
fix(ivy): deduplicate imported module in r3_injector

### PR DESCRIPTION
**Pull request by ocombe**
## PR Type
What kind of change does this PR introduce?

- [x] Bugfix

## What is the current behavior?
The r3_injector doesn't deduplicate imported modules, meaning that you can import the same module multiple times and the providers will be overwritten everytime

## What is the new behavior?
We deduplicate imports, only the first one matters like it is the case with the current renderer.
This also enables the tests from platform-server for ivy.

## Does this PR introduce a breaking change?

- [x] No

### CODE REVIEW

File: packages/core/src/di/r3_injector.ts [hunk 1]
```diff
@@ -251,8 +257,11 @@ export class R3Injector {
       // Before processing defType's imports, add it to the set of parents. This way, if it ends
       // up deeply importing itself, this can be detected.
       parents.add(defType);
+      // Add it to the set of dedups. This way we can detect multiple imports of the same module
+      dedup.add(defType);
+      // console.log('imports', def.imports);
```
Comment (alfaproject) [inline packages/core/src/di/r3_injector.ts]: I guess you can remove this now

File: packages/core/src/di/r3_injector.ts [hunk 2]
```diff
@@ -251,8 +257,11 @@ export class R3Injector {
       // Before processing defType's imports, add it to the set of parents. This way, if it ends
       // up deeply importing itself, this can be detected.
       parents.add(defType);
+      // Add it to the set of dedups. This way we can detect multiple imports of the same module
+      dedup.add(defType);
```
Comment (alfaproject) [inline packages/core/src/di/r3_injector.ts]: `parents` seem to be the exact same set. Maybe can just use that?

Comment (ocombe) [reply packages/core/src/di/r3_injector.ts]: yeah, actually it doesn't check for duplicate dependencies, I need to change my fix

File: packages/core/src/di/r3_injector.ts [hunk 3]
```diff
@@ -218,6 +219,12 @@ export class R3Injector {
     const defType: InjectorType<any> =
         (ngModule === undefined) ? (defOrWrappedDef as InjectorType<any>) : ngModule;
 
+    let isDuplicate = false;
```
[the remark this answers is not in the document]
Comment (ocombe) [reply packages/core/src/di/r3_injector.ts]: so that circular dependencies can be detected

[the remark this answers is not in the document]
Comment (JoostK) [reply packages/core/src/di/r3_injector.ts]: @ocombe Ah, right! Could use a comment here 👍

**Review by alxhub — APPROVED** [review]
The only thing I would suggest is splitting this into two commits, where one implements the changes to `R3Injector` and the other enables the tests.

File: packages/core/src/di/r3_injector.ts [hunk 4]
```diff
@@ -218,6 +220,16 @@ export class R3Injector {
     const defType: InjectorType<any> =
         (ngModule === undefined) ? (defOrWrappedDef as InjectorType<any>) : ngModule;
 
+    // Check for circular dependencies.
+    if (parents.has(defType)) {
+      throw new Error(`Circular dependency: type ${stringify(defType)} ends up importing itself.`);
```
Comment (IgorMinar) [inline packages/core/src/di/r3_injector.ts]: Can you update this to a better error message: `Circular dependency in DI: type ${stringify(defType)} (indirectly) depends on itself.`

Comment (IgorMinar) [reply packages/core/src/di/r3_injector.ts]: Actually, we should implement this as stack and print out the dependency path for better debugging. so the error would be:

`Circular dependency in DI detected for type ${stringify(defType)}. Dependency path: Foo > Bar > Baz > Foo`

File: packages/core/src/di/r3_injector.ts [hunk 5]
```diff
@@ -108,8 +108,10 @@ export class R3Injector {
       readonly parent: Injector) {
     // Start off by creating Records for every provider declared in every InjectorType
     // included transitively in `def`.
+    const dedupSet = new Set<InjectorType<any>>();
```
Comment (IgorMinar) [inline packages/core/src/di/r3_injector.ts]: can you implement this as a stack rather than set and turn on the check only in dev mode?

File: packages/platform-server/test/BUILD.bazel [hunk 6]
```diff
@@ -17,6 +17,7 @@ ts_library(
         "//packages/http/testing",
         "//packages/platform-browser",
         "//packages/platform-server",
+        "//packages/private/testing",
```
Comment (IgorMinar) [inline packages/platform-server/test/BUILD.bazel]: all these fixmeIvy changes are unrelated to the original change. Please move them to a separate commit.

**Review by IgorMinar — CHANGES_REQUESTED** [review]

**Review by IgorMinar — APPROVED** [review]
thanks for making the changes.

**Comment by mhevery** [comment]
http://test/OCL:222084214:BASE:222084226:1542646783789:bb3b4d6d
